design/authenticators/authn_sut/authn_sut_solution_design.md

```diff
@@ -0,0 +1,384 @@
+
+The SUT will be a cryptographically secure, random string. It will be generated
+by the Conjur server and stored in the database. The SUT will be associated with
+the user and expiration time. The SUT will be deleted once it is utilized
+to authenticate the user to Conjur.
+
+When we store the SUT in the database, we will store a salted hash of the SUT instead
+of the SUT itself. This will prevent an attacker with access to the database
+from being able to use the SUTs to authenticate to Conjur. The SUT will be
+salted and hashed using a cryptographically secure hash function, such as SHA-256.
+When a user presents a SUT to authenticate to Conjur, we will hash the SUT and
+compare it to the hash stored in the database.
+
+##### SUT Format
+
+We cannot use JWTs for the SUT as we need the ability to revoke the SUT after it
+is used, a feature not easily implemented with JWTs. This would make it
+impossible to prevent replay attacks.
+
+##### SUT Expiration and Revocation
+
+The SUT will have a short expiration time, on the order of 20-30 seconds. This is
+enough time to allow the client to use the SUT to authenticate to Conjur, even
+in the case of network latency, but short enough to limit the window of
+opportunity for an attacker to steal and use the SUT.
+
+Additionally, only one SUT will be valid at a time for a given user. If a new
+SUT is generated for a user, the previous SUT will be deleted.
+
+When a SUT is used (or revoked), it will be deleted from the database.
+Expired SUTs can also safely be deleted from the database, though
+this is not strictly necessary since they will not be valid past
+their expiration time.
+When deleting used or expired SUTs, we must not cascade the delete
+since this would delete any related audit logs.
+
+##### Code Verifier and Code Challenge
+
+As an additional layer of security, the client will generate a code verifier,
+which can be a Guid or other cryptographically secure random string. The code
+will then be hashed and sent along with the login request to be associated
+with the generated SUT. This hashed code verifier, known as the code
+challenge, will be stored by Conjur along with the SUT. When the client
+attempts to authenticate with the SUT, they will also send the code verifier
+(in plain text form), and Conjur will hash it and compare it to the stored
+code challenge.
+
+Initially we will support the SHA-256 algorithm for the code challenge. In the
+future we may want to support additional algorithms.
+
+The code challenge and code verifier flow are modeled after OIDC PKCE.
+
+##### Open Questions
+
+- Is it possible to restrict use of the SUT to a specific IP range (like with
+  hostfactory tokens) or other criteria to limit risk if the SUT is leaked?
+- Should we provide a way for clients to register a key pair to use for
+  signing the SUT generation request? This would allow us to verify that the
+  request came from a trusted client.
+- Should we create a job to clean up expired SUTs?
```

Comment:
Absolutely.  We can use Cron as it's already present in Conjur and Conjur Enterprise.

Reply:
Where is Cron used in Conjur?

Reply:
My mistake. Cron is part of Conjur Enterprise, not Conjur. Two options that come to mind are:

- Remove expired single-use tokens during creation, exchange, or both.
- Create a service like the variable rotation service to run and periodically remove expired tokens

Reply:
OK got it. For now I'll update this doc to allow for some extra points for figuring that out during the implementation.